application/common/models/Method.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+namespace common\models;
+
+use common\exceptions\InvalidCodeException;
+use common\helpers\MySqlDateTime;
+use common\helpers\Utils;
+use yii\helpers\ArrayHelper;
+use yii\web\ConflictHttpException;
+use yii\web\ServerErrorHttpException;
+
+/**
+ * Class Method
+ * @package common\models
+ *
+ */
+class Method extends MethodBase
+{
+    public function rules()
+    {
+        return ArrayHelper::merge(
+            [
+                [
+                    ['uid'], 'default', 'value' => Utils::generateRandomString(),
+                ],
+
+                [
+                    ['verified', 'verification_attempts'], 'default', 'value' => 0,
+                ],
+
+                [
+                    'verification_code', 'default', 'when' => function() { return $this->getIsNewRecord(); },
+                    'value' => Utils::getRandomDigits(\Yii::$app->params['reset']['codeLength']),
+                ],
+
+                [
+                    'verification_expires', 'default', 'when' => function() { return $this->getIsNewRecord(); },
+                    'value' => MySqlDateTime::formatDateTime(time() + \Yii::$app->params['reset']['lifetimeSeconds']),
+                ],
+
+                [
+                    'value', 'email',
+                ],
+
+                [
+                    ['created'], 'default', 'value' => MySqlDateTime::now(),
+                ],
+
+            ],
+            parent::rules()
+        );
+    }
+
+    public function fields()
+    {
+        return [
+            'id' => function() { return $this->uid; },
+            'value',
+        ];
+    }
+
+    /**
+     * @return string
+     * @throws \Exception
+     */
+    public function getMaskedValue()
+    {
+        return Utils::maskEmail($this->value);
+    }
+
+    public function isVerified()
+    {
+        return $this->verified === 1 ? true : false;
+    }
+
+    /**
+     * Send verification message
+     * @throws ConflictHttpException
+     * @throws ServerErrorHttpException
+     */
+    public function sendVerification()
+    {
+        if ($this->isVerified()) {
+            throw new ConflictHttpException('Method already verified', 1540689804);
+        }
+
+        $this->verification_attempts++;
+        if ( ! $this->save()) {
+            throw new ServerErrorHttpException('Save error after incrementing attempts', 1461441850);
+        }
+
+//        $this->sendVerificationEmail();
+    }
+
+    /**
+     * Send a verification email message
+     */
+    public function sendVerificationEmail()
+    {
+        Verification::sendEmail(
```

Comment:
The `Verification` class doesn't seem to exist.

Reply:
Yup. TBD.